### PR TITLE
fix: enable magic HTML routes for `module` output

### PR DESCRIPTION
**Pull request by snitin315**
- [x] This is a **bugfix**
- [ ] This is a **feature**
- [ ] This is a **code refactor**
- [x] This is a **test update**
- [x] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?
Yes

### Motivation / Use-Case

enable magic HTML routes for `.mjs` output.

### Breaking Changes

None

### Additional Info
No

### CODE REVIEW

File: lib/Server.js [hunk 1]
```diff
@@ -1931,7 +1931,29 @@ class Server {
       const _path = req.path;
 
       try {
-        const filename = this.middleware.getFilenameFromUrl(`${_path}.js`);
+        const compilerOptions = this.getCompilerOptions();
+        const isOutputModuleEnabled =
+          compilerOptions.experiments &&
+          compilerOptions.experiments.outputModule;
```
Comment (alexander-akait) [inline lib/Server.js:1937]: Developer can set `js` extension for modules too, ideally we should look at info for assets, i.e. compilation.getAssetInfo and check on `module: true | false | undefined`

File: lib/Server.js [hunk 2]
```diff
@@ -1931,7 +1931,49 @@ class Server {
...
+          // If key = main.other.mjs && filename = main => false
+          // If key = main.other.mjs && filename = main.other => true
+          if (
+            splitKey.length - 1 === splitOutputFilename.length &&
+            splitKey[0] === outputFilename.split(".")[0]
+          ) {
+            extension = splitKey[splitKey.length - 1];
+            isOutputModule = value.javascriptModule;
+          }
+        });
```
Comment (alexander-akait) [inline lib/Server.js:1962]: We have `path.ext()` here

Comment (snitin315) [reply lib/Server.js:1962]: I will update 👍🏻

**Review by alexander-akait — COMMENTED** [review]
Can we add test for multi compiler mode?

**Comment by snitin315** [comment]
Something wrong with Node 12 and MacOS, looking into it.

**Comment by alexander-akait** [comment]
@snitin315 let's finish

**Comment by snitin315** [comment]
It seems https://github.com/webpack/webpack-dev-server/pull/3840 broke the magic HTML feature for multi-compiler mode.

/cc @alexander-akait

**Comment by alexander-akait** [comment]
Can you provide example?

**Comment by snitin315** [comment]
- cd examples/magic-html
- update `webpack.config.js` as
```diff
- module.exports = setup({
+ module.exports = [setup({
  context: __dirname,
  entry: "./app.js",
  devServer: {
    magicHtml: true,
  },
- });
+ })];
```
- `npx webpack serve --open`
- go to `localhost:8080/main` -> `Can't get main`

**Comment by snitin315** [comment]
I think it is specific to this PR only, I am looking into it.

File: lib/Server.js [hunk 3]
```diff
@@ -1974,10 +2016,11 @@ class Server {
 
         // Serve a page that executes the javascript
         const queries = req._parsedUrl.search || "";
-        const responsePage = `<!DOCTYPE html><html><head><meta charset="utf-8"/></head><body><script type="text/javascript" charset="utf-8" src="${_path}.js${queries}"></script></body></html>`;
+        const responsePage = `<!DOCTYPE html><html><head><meta charset="utf-8"/></head><body><script type="${scriptType}" charset="utf-8" src="${_path}${extension}${queries}"></script></body></html>`;
 
         res.send(responsePage);
       } catch (error) {
+        console.log(error);
```
Comment (snitin315) [inline lib/Server.js:2023]:
```suggestion
```

**Review by alexander-akait — COMMENTED** [review]
Please add test for `publicPath`

File: lib/Server.js [hunk 4]
```diff
@@ -1963,7 +1963,49 @@ class Server {
...
+        let scriptType;
+
+        // _path = "/main" -> outputFilename = "main"
+        // _path = "/main.other" -> outputFilename = "main.other"
+        const outputFilename = _path.slice(1);
+
+        let isOutputModule = false;
+
+        if (typeof this.compiler.compilers !== "undefined") {
+          this.stats.stats.forEach((stats) => {
```
Comment (alexander-akait) [inline lib/Server.js:1977]: Also lets improve this using and avoid if/else:
```
const stats = this.stats.stats ? this.stats.stats : [this.stats];
```

File: lib/Server.js [hunk 5]
```diff
@@ -1963,7 +1963,49 @@ class Server {
...
+            stats.compilation.assetsInfo.forEach((value, key) => {
+              const assetExtension = path.extname(key);
+              const assetBasename = path.basename(key, assetExtension);
+
+              if (assetBasename === outputFilename) {
+                extension = assetExtension;
+                isOutputModule = value.javascriptModule;
+              }
+            });
+          });
```
Comment (alexander-akait) [inline lib/Server.js:1987]: Using `forEach` is slow, better to move this in `const findInfo = () => { for (const item of stats) { for (const assetsInfo of item.compilation.assetsInfo) { if (one === two) {  return { extension, isModule } } } }}` (pseudo code)

**Comment by snitin315** [comment]
> Please add test for publicPath

`output.publicPath` or `static.publicPath`. Also, what should be the expected behavior?

**Comment by snitin315** [comment]
removed in v5

**Pull request closed without merging**